### PR TITLE
APPS-2392 - Redesign APIs for authentication, persistence manager, etc

### PR DESCRIPTION
**Pull request by kcoleman731**
* Implements multiple protocols for Manager and Model classes to make the Atlas Messenger easily customizable.

### CODE REVIEW

File: Atlas Messenger.xcworkspace/xcshareddata/Atlas Messenger.xcscmblueprint [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositorySystemKey" : "com.apple.dt.Xcode.sourcecontrol.Git",
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryIdentifierKey" : "555B8AD8B8A34913DE837AE8E32AA22302CA6BED"
+    },
+    {
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:layerhq\/Layer-Messenger-iOS.git",
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositorySystemKey" : "com.apple.dt.Xcode.sourcecontrol.Git",
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryIdentifierKey" : "704F552CEB6415A9F52EE9446435355DF9848168"
+    }
+  ]
+}
```
Comment (blakewatters) [inline Atlas Messenger.xcworkspace/xcshareddata/Atlas Messenger.xcscmblueprint]: We don't want this in Git do we?

File: Code/ATLMAppDelegate.m [hunk 2]
```diff
@@ -81,6 +86,33 @@ - (void)applicationWillResignActive:(UIApplication *)application
 
 #pragma mark - Setup
 
+- (void)setupLayerClient
+{
+    NSString *appID = ATLMLayerAppID ?: [[NSUserDefaults standardUserDefaults] valueForKey:ATLMLayerApplicationID];
```
Comment (blakewatters) [inline Code/ATLMAppDelegate.m]: I'd probably pass this in as an argument from the entry point so you don't have configuration logic spread across instance methods

File: Code/ATLMAppDelegate.m [hunk 3]
```diff
@@ -81,6 +86,33 @@ - (void)applicationWillResignActive:(UIApplication *)application
...
+    if (appID) {
+        if (!self.layerClient) {
```
Comment (blakewatters) [inline Code/ATLMAppDelegate.m]: Why are we guarding here? What's the case where this is called multiple times?

File: Code/ATLMAppDelegate.m [hunk 4]
```diff
@@ -266,11 +264,25 @@ - (void)application:(UIApplication *)application handleActionWithIdentifier:(nul
...
+{
+    NSURL *conversationIdentifier = [NSURL URLWithString:[remoteNotification valueForKeyPath:@"layer.conversation_identifier"]];
+    return [self.applicationController.layerClient existingConversationForIdentifier:conversationIdentifier];
+}
+
+- (void)navigateToViewForConversation:(LYRConversation *)conversation
+{
+    if (![NSThread isMainThread]) {
+        @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"Attempted to navigate UI from non-main thread" userInfo:nil];
+    }
```
Comment (blakewatters) [inline Code/ATLMAppDelegate.m]: You could always dispatch async to the main thread to make this a non-issue

File: Code/ATLMAuthenticating.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+//  Copyright © 2016 Layer, Inc. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <Atlas/Atlas.h>
+
+/**
+ @abstract The `ATLMATLMAuthenticating` protocol must be adopted by objects that model users in the Atlas Messenger application.
+ */
+@protocol ATLMAuthenticating <NSObject, NSCoding, ATLParticipant>
```
Comment (blakewatters) [inline Code/ATLMAuthenticating.h]: I thought `ATLParticipant` went away?

File: Code/ATLMAuthenticating.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ @abstract The password for the user.
+ */
+@property (nonatomic, readonly) NSString *password;
+
+/**
+ @abstract The password confirmation for the user.
+ */
+@property (nonatomic, readonly) NSString *passwordConfirmation;
```
Comment (blakewatters) [inline Code/ATLMAuthenticating.h]: eh mandating email, password and confirmation isn't super flexible

File: Code/ATLMPersistenceManaging.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+//  Copyright © 2016 Layer, Inc. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import "ATLMSession.h"
+
+/**
+ @abstract The `ATLMPersistenceManaging` protocol must be adopted by object wishing to perform persitence operations in the Atlas Messenger Application.
+ */
+@protocol ATLMPersistenceManaging <NSObject>
```
Comment (blakewatters) [inline Code/ATLMPersistenceManaging.h]: Do we even need this protocol? It seems like the difference between a transient and a persistent session can be abstracted away

File: Code/ATLMAuthenticationProvider.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ @abstract The initializer for the `ATLMAuthenticationProvider`.
+ @param baseURL The base url for the Layer Identity provider.
+ */
++ (nonnull instancetype)providerWithBaseURL:(nonnull NSURL *)baseURL;
+
+/**
+ @abstract Updates the `ATLMAuthenticationProvider` with a Layer app ID.
+ @param appID The Layer app ID.
+ */
+- (void)updateWithAppID:(nonnull NSURL *)appID;
```
Comment (layer-engineering) [inline Code/ATLMAuthenticationProvider.h]: Figure out how to eliminate it.

File: Code/ATLMAppDelegate.m [hunk 9]
```diff
@@ -52,35 +53,57 @@ @implementation ATLMAppDelegate
...
+    NSString *appID = ATLMLayerAppID ?: [[NSUserDefaults standardUserDefaults] valueForKey:ATLMLayerApplicationID];
+    if (!appID) {
+        return;
+    }
+    if (!self.layerClient) {
+        NSDictionary *options = @{LYRClientOptionSynchronizationPolicy : @(LYRClientSynchronizationPolicyMessageCount), LYRClientOptionSynchronizationMessageCount: @(10)};
+        self.layerClient = [ATLMLayerClient clientWithAppID:[NSURL URLWithString:appID] options:options];
+        self.layerClient.autodownloadMIMETypes = [NSSet setWithObjects:ATLMIMETypeImageJPEGPreview, ATLMIMETypeTextPlain, nil];
+        [self.applicationController updateWithLayerClient:self.layerClient];
+        [(ATLMAuthenticationProvider *)self.applicationController.authenticationProvider updateWithAppID:[NSURL URLWithString:appID]];
```
Comment (blakewatters) [inline Code/ATLMAppDelegate.m]: What's up with this cast?

File: Code/ATLMAppDelegate.m [hunk 10]
```diff
@@ -266,11 +240,28 @@ - (void)application:(UIApplication *)application handleActionWithIdentifier:(nul
     completionHandler();
 }
 
+- (LYRConversation *)conversationFromRemoteNotification:(NSDictionary *)remoteNotification
+{
+    NSURL *conversationIdentifier = [NSURL URLWithString:[remoteNotification valueForKeyPath:@"layer.conversation_identifier"]];
+    return [(ATLMLayerClient *)self.applicationController.layerClient existingConversationForIdentifier:conversationIdentifier];
```
Comment (blakewatters) [inline Code/ATLMAppDelegate.m]: What's up with this cast?

File: Code/ATLMAppDelegate.m [hunk 11]
```diff
@@ -396,7 +371,7 @@ - (void)configureGlobalUserInterfaceAttributes
 
 - (void)setApplicationBadgeNumber
 {
-    NSUInteger countOfUnreadMessages = [self.applicationController.layerClient countOfUnreadMessages];
+    NSUInteger countOfUnreadMessages = [(ATLMLayerClient *)self.applicationController.layerClient countOfUnreadMessages];
```
Comment (blakewatters) [inline Code/ATLMAppDelegate.m]: Cast?

File: Code/ATLMAuthenticationProvider.h [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+@end
+
+
```
Comment (blakewatters) [inline Code/ATLMAuthenticationProvider.h]: Extra whitespace

File: Code/ATLMAuthenticationProvider.h [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+/*
+ @abstract A key whose value should be the last name of an authenticating user.
+ */
+extern NSString * _Nonnull const ATLMLastNameKey;
+
+/**
+ @abstract The `ATLMAuthenticationProvider` conforms to the `ATLMAuthenticating` protocol. It provides for making requests to the Layer Identity Provider in order to request identity tokens needed of LayerKit authentication.
+ */
+@interface ATLMAuthenticationProvider : NSObject <ATLMAuthenticating>
```
Comment (blakewatters) [inline Code/ATLMAuthenticationProvider.h]: Let's do a better name that is more descriptive

File: Code/ATLMAuthenticationProvider.m [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+//
+
+
+#import "ATLMAuthenticationProvider.h"
+#import "ATLMHTTPResponseSerializer.h"
+#import "ATLMConstants.h"
+
+NSString *const ATLMFirstNameKey = @"ATLMFirstNameKey";
+NSString *const ATLMLastNameKey = @"ATLMLastNameKey";
+NSString *const ATLMCredentialsKey = @"ATLMCredentialsKey";
```
Comment (blakewatters) [inline Code/ATLMAuthenticationProvider.m]: Not sure why these needs to be constants... is there a corresponding login screen?

File: Code/ATLMAuthenticationProvider.m [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+- (void)updateWithAppID:(NSURL *)appID
+{
+    _appID = appID;
+    _URLSession = [self defaultURLSession];
+}
+
+- (NSURLSession *)defaultURLSession
+{
+    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration ephemeralSessionConfiguration];
+    configuration.HTTPAdditionalHeaders = @{ @"Accept": @"application/json", @"X_LAYER_APP_ID": self.appID.absoluteString };
```
Comment (blakewatters) [inline Code/ATLMAuthenticationProvider.m]: Formatting

File: Code/ATLMAuthenticationProvider.m [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+- (void)authenticateWithCredentials:(NSDictionary *)credentials nonce:(NSString *)nonce completion:(void (^)(NSString *identityToken, NSError *error))completion
+{
+    NSString *firstName = credentials[ATLMFirstNameKey];
+    NSString *lastName = credentials[ATLMLastNameKey];
+    NSString *displayName = [NSString stringWithFormat:@"%@ %@", firstName, lastName];
+    NSString *appUUID = [[self.appID pathComponents] lastObject];
+    NSString *urlString = [NSString stringWithFormat:@"apps/%@/atlas_identities", appUUID];
+    NSURL *URL = [NSURL URLWithString:urlString relativeToURL:self.baseURL];
+    NSDictionary *parameters = @{ @"user" : @{ @"first_name": firstName, @"last_name": lastName, @"display_name": displayName, @"avatar_url": @"" }, @"nonce" : nonce };
```
Comment (blakewatters) [inline Code/ATLMAuthenticationProvider.m]: Formatting

File: Code/ATLMAuthenticationProvider.m [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        NSDictionary *userDetails;
+        BOOL success = [ATLMHTTPResponseSerializer responseObject:&userDetails withData:data response:(NSHTTPURLResponse *)response error:&serializationError];
+        if (success) {
+            dispatch_async(dispatch_get_main_queue(), ^{
+                NSLog(@"User JSON: %@", userDetails);
+                NSString *identityToken = userDetails[ATLMAtlasIdentityTokenKey];
+                completion(identityToken, nil);
+            });
+            [[NSUserDefaults standardUserDefaults] setValue:credentials forKey:ATLMCredentialsKey];
+            [[NSUserDefaults standardUserDefaults] synchronize];
```
Comment (blakewatters) [inline Code/ATLMAuthenticationProvider.m]: Whoa. We don't want to store passwords into user defaults.

File: Code/ATLMAuthenticationProvider.m [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+        }
+    }] resume];
+}
+
+- (void)refreshAuthenticationWithNonce:(NSString *)nonce completion:(void (^)(NSString *identityToken, NSError *error))completion
+{
+    NSDictionary *credentails = [[NSUserDefaults standardUserDefaults] objectForKey:ATLMCredentialsKey];
+    [self authenticateWithCredentials:credentails nonce:nonce completion:^(NSString * _Nonnull identityToken, NSError * _Nonnull error) {
+        completion(identityToken, error);
+    }];
```
Comment (blakewatters) [inline Code/ATLMAuthenticationProvider.m]: I think here we just want to use the existing cookied session to fetch a new identity token

File: Code/Controllers/ATLMApplicationController.h [hunk 19]
```diff
@@ -41,30 +41,42 @@ extern NSString *const ATLMConversationDeletedNotification;
...
+ @param credentials An `NSDictionary` containing authetication credentials. 
+ @param completions A block to be called upon completion of the operation.
+ */
+- (void)authenticateWithCredentials:(nonnull NSDictionary *)credentials completion:(nonnull void (^)(LYRSession * _Nonnull session, NSError * _Nonnull error))completion;
+
+/**
+ @abstract Updates the controller with a `LYRClient` instance.
+ @param client The `LYRClient` instancce.
+ */
+- (void)updateWithLayerClient:(nonnull LYRClient *)client;
```
Comment (blakewatters) [inline Code/Controllers/ATLMApplicationController.h]: Why do we need this?

File: Code/Controllers/ATLMConversationDetailViewController.m [hunk 20]
```diff
@@ -444,7 +444,7 @@ - (void)participantTableViewController:(ATLParticipantTableViewController *)part
 - (void)switchToConversationForParticipants
 {
     NSSet *participantIdentifiers = [self.participants valueForKey:@"userID"];
-    LYRConversation *conversation = [self.applicationController.layerClient existingConversationForParticipants:participantIdentifiers];
+    LYRConversation *conversation = [(ATLMLayerClient *)self.applicationController.layerClient existingConversationForParticipants:participantIdentifiers];
```
Comment (blakewatters) [inline Code/Controllers/ATLMConversationDetailViewController.m]: Cast

File: Code/Utilities/ATLMConstants.h [hunk 21]
```diff
@@ -20,6 +20,14 @@
 
 #import <UIKit/UIKit.h>
 
+extern NSString *const ATLMUserDidAuthenticateNotification;
+extern NSString *const ATLMUserDidDeauthenticateNotification;
+extern NSString *const ATLMApplicationDidSynchronizeParticipants;
```
Comment (blakewatters) [inline Code/Utilities/ATLMConstants.h]: Do we need this?

File: Code/Utilities/ATLMConstants.m [hunk 22]
```diff
@@ -8,6 +8,14 @@
 
 #import "ATLMConstants.h"
 
+NSString *const ATLMUserDidAuthenticateNotification = @"ATLMUserDidAuthenticateNotification";
+NSString *const ATLMUserDidDeauthenticateNotification = @"ATLMUserDidDeauthenticateNotification";
+NSString *const ATLMApplicationDidSynchronizeParticipants = @"ATLMApplicationDidSynchronizeParticipants";
```
Comment (blakewatters) [inline Code/Utilities/ATLMConstants.m]: Delete this

File: Code/Controllers/ATLMApplicationController.m [hunk 23]
```diff
@@ -53,28 +53,43 @@ - (void)dealloc
...
 
 - (void)layerClient:(LYRClient *)client didReceiveAuthenticationChallengeWithNonce:(NSString *)nonce
 {
-    NSLog(@"Layer Client did recieve authentication challenge with nonce: %@", nonce);
-    ATLMUser *user = self.APIManager.authenticatedSession.user;
-    if (!user) return;
-    //TODO - Handle Auth Challenge;
+    [self.authenticationProvider refreshAuthenticationWithNonce:nonce completion:^(NSString * _Nonnull identityToken, NSError * _Nonnull error) {
+        //
+    }];
```
Comment (layer-engineering) [inline Code/Controllers/ATLMApplicationController.m]: yick

File: Code/ATLMAppDelegate.m [hunk 24]
```diff
@@ -20,159 +20,88 @@
...
-            NSLog(@"Layer Client Connected");
-        }];
-    }
+    NSParameterAssert(appID);
+    ATLMAuthenticationProvider *authenticationProvider = [ATLMAuthenticationProvider providerWithBaseURL:ATLMRailsBaseURL(ATLMEnvironmentProduction) layerAppID:appID];
+    
+    // Configure the Layer Client options.
+    LYRClientOptions *clientOptions = [LYRClientOptions new];
+    clientOptions.synchronizationPolicy = LYRClientSynchronizationPolicyPartialHistory;
+    clientOptions.partialHistoryMessageCount = 20;
```
Comment (chipxsd) [inline Code/ATLMAppDelegate.m]: who's in charge of setting the autodownload mimetypes?

**Comment by blakewatters** [comment]
Squashed and merged into master

**Pull request closed without merging**